src/connector.rs

```diff
@@ -65,165 +85,184 @@ impl EthConnectorContract {
 
-        let promise1 = sdk::promise_then(
-            promise0,
-            &account_id,
-            b"finish_deposit_eth",
-            &data[..],
-            NO_DEPOSIT,
-            GAS_FOR_FINISH_DEPOSIT,
-        );
+        // Finalize deposit
+        let promise1 = match self.parse_event_message(&event.recipient) {
+            // Deposit to NEAR accounts
+            TokenMessageData::Near(account_id) => {
+                let predecessor_account_id =
+                    String::from_utf8(sdk::predecessor_account_id()).unwrap();
+                let data = FinishDepositCallArgs {
+                    new_owner_id: account_id,
+                    amount: event.amount.as_u128(),
+                    proof,
+                    relayer_id: Some(predecessor_account_id),
+                    fee: Some(event.fee.as_u128()),
+                }
+                .try_to_vec()
+                .unwrap();
+
+                sdk::promise_then(
+                    promise0,
+                    &sdk::current_account_id(),
+                    b"finish_deposit_near",
+                    &data[..],
+                    NO_DEPOSIT,
+                    GAS_FOR_FINISH_DEPOSIT,
+                )
+            }
+            // Deposit to Eth/ERC20 accounts
+            // fee mint in the `ft_on_transfer` callback method
+            TokenMessageData::Eth { address, message } => {
+                let current_account_id = String::from_utf8(sdk::current_account_id()).unwrap();
+                // Send to self - current account id
+                let data = FinishDepositCallArgs {
+                    new_owner_id: current_account_id,
+                    amount: event.amount.as_u128(),
+                    proof,
+                    relayer_id: None,
+                    fee: None,
+                }
+                .try_to_vec()
+                .unwrap();
+
+                let internal_promise = sdk::promise_then(
+                    promise0,
+                    &sdk::current_account_id(),
+                    b"finish_deposit_near",
+                    &data[..],
+                    NO_DEPOSIT,
+                    GAS_FOR_FINISH_DEPOSIT,
+                );
+                // Transfer to self and then transfer ETH in `ft_on_transfer`
+                // address - is NEAR account
+                let transfer_data = TransferCallCallArgs {
+                    receiver_id: address,
+                    amount: event.amount.as_u128(),
+                    memo: None,
+                    msg: self.set_message_for_on_transfer(event.fee, message),
+                }
+                .try_to_vec()
+                .unwrap();
+                sdk::promise_then(
+                    internal_promise,
+                    &sdk::current_account_id(),
+                    b"ft_transfer_call",
+                    &transfer_data[..],
+                    1,
+                    GAS_FOR_TRANSFER_CALL,
+                )
+            }
```

Comment:
There's no check if the `internal_promise` was successfully completed, so even if `finalise_deposit_near()` failed let's say on `ERR_VERIFY_PROOF`, the `ft_transfer_call()` will be executed anyway.